automated-testing/container-testing/azdo-container-dev-test-release.md

```diff
@@ -0,0 +1,286 @@
+# Building Containers with Azure DevOps using DevTest Pattern
+
+In this documents, we highlight learnings from applying the DevTest pattern to container development in Azure DevOps through pipelines.
+
+The pattern enabled as to build container for development, testing and releasing the container for further reuse (production ready).
+
+We will dive into tools needed to build, test and push a container, our environment and go through each step separately.
+
+Follow this link to dive deeper or revisit the [DevTest pattern](https://docs.microsoft.com/en-us/azure/architecture/solution-ideas/articles/dev-test-paas).
+
+## Table of Contents
+
+[Build the Container](#Build-The-Container)
+[Test the Container](#Test-the-Container)
+[Push Container](#Push-container)
+[References](#References)
+
+## Build the Container
+
+The first step in container development, after creating the necessary Dockerfiles and source code, is building the container. Even the Dockerfile itself can include some basic testing. Code tests are performed when pushing the code to the repository origin, where it is then used to build the container.
+
+The first step in our pipeline is to run the `docker build` command with a temporary tag and the required build arguments:
+
+```yaml
+- task: Bash@3
+  name: BuildImage
+  displayName: 'Build the image via docker'
+  inputs:
+    workingDirectory: "$(System.DefaultWorkingDirectory)${{ parameters.buildDirectory }}"
+    targetType: 'inline'
+    script: |
+      docker build -t ${{ parameters.imageName }} --build-arg YOUR_BUILD_ARG -f ${{ parameters.dockerfileName }} .
+  env:
+    PredefinedPassword: $(Password)
+    NewVariable: "newVariableValue"
+```
+
+This task includes the parameters `buildDirectory`, `imageName` and `dockerfileName`, which have to be set beforehand.
+This task can for example be used in a template for multiple containers to improve code reuse.
+
+It is also possible to pass environment variables directly to the Dockerfile through the `env` section of the task.
+
+If this task succeeds, the Dockerfile was build without errors and we can continue to testing the container itself.
+
+## Test the Container
+
+To test the container, we are using the tox environment.
```

Comment:
Can you add some context on tox and how it fits in this doc, the link can be left there for those that want to dig deeper.